File: VerticalResolution/CPPVersion/resolutionir.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <numeric>
#include <functional>
#include <complex>
#include <algorithm>
using namespace std;



double stddev(std::vector<double> v)
{
	if (v.size()  <= 2)
		return 0.;
	double sum = std::accumulate(v.begin(), v.end(), 0.0);
	double m =  sum / v.size();

	double accum = 0.0;
	for(std::vector<double>::iterator d = v.begin(); d != v.end(); ++d)
	{
		accum += (*d - m) * (*d - m);
	}

	double stdev = sqrt(accum / (v.size()-1));
	return stdev;
}

void convolve(const vector<double>& vIrc, const vector<double>& vCo, vector<double> & ir)
{
	ir.resize(vIrc.size(), 0.);
	int nc = (vCo.size() - 1) / 2;
	for(unsigned j = nc; j < vIrc.size() - nc; ++j)
	{
		double tmp = 0;
		for(int a = - nc; a < nc + 1; ++a)
		{
			tmp += vIrc[j + a] * vCo[nc + a];
		}
		ir[j] = tmp;
	}
}


void NDAAC_ResolIR(double vZsampling, const vector<double>& vCoef, vector<double> vIRinp, unsigned &rStatus, double &rDzfwhm, vector<double>& rIRout)
{

	unsigned ncoef = vCoef.size();
	unsigned nk = vIRinp.size();
	rStatus = 0;
	double missval = -99;
	vector<double> irout(vIRinp.begin(), vIRinp.end());
	vector<double> irconvol(vIRinp.begin(), vIRinp.end());
	vector<double> ir(nk, 0.);
	int nc = (ncoef - 1) / 2;
	if((ncoef % 2) == 0)
	{
		cout<<"ERROR: the number of coefficients is even"<<endl;
		return;
	}
	if ((ncoef == 1) and vCoef[0] != 1)
	{
		rStatus = 0;
		cout<<"ERROR: we have a non-filter with a non unit coefficient"<<endl;
		return;
	}
	if ((ncoef == 1) and vCoef[0] == 1)
	{
		rStatus = 1;
	}else
	{
		rStatus = 42;
		for(int ic = 1; ic < nc + 1; ++ic)
		{
			if(abs(vCoef[nc - ic]) < 1E-20 && abs(vCoef[nc + ic]) < 1E-20)
			{
				cout<<"zero  "<<endl;
			}else if (abs(vCoef[nc - ic] - vCoef[nc + ic]) < 1E-3 * abs(vCoef[nc + ic]))
			{
				if (rStatus == 2)
				{
					rStatus = 0;
					cout<<"ERROR: we have a mix of symmetric and antisymmetric"<<endl;
					return;
				}
				rStatus = 1;
			}else if (abs(vCoef[nc - ic] + vCoef[nc + ic]) < 1E-3 * abs(vCoef[nc + ic]) and vCoef[nc] == 0)
			{
				if (rStatus == 1)
				{
					rStatus = 0;
					cout<<"ERROR: we have a mix of symmetric and antisymmetric"<<endl;
					return;
				}
				rStatus = 2;
			}else
			{
				rStatus = 0;
				cout<<"ERROR: we have a mix of symmetric and antisymmetric (impossible parameters)"<<endl;
				return;
			}

		}	
	}
	if (accumulate(irconvol.begin(), irconvol.end(),0.) == 0 and stddev(irconvol) == 0)
	{
		irconvol[nk / 2] = 1;
	}
	if (rStatus == 2)
	{
		cout<<"Status 2"<<endl;
		vector<double> irtemp(irconvol.size(), 0.);
		irtemp[0] =irconvol[0];
		double maxirtemp = irtemp[0];
		for(unsigned k=0; k < nk; ++k)
		{
			//irtemp[k] = sum(irconvol[0:k + 1]);
			irtemp[k] = accumulate(irconvol.begin(), irconvol.begin() + k + 1, 0); 
			if (irtemp[k] > maxirtemp)
			{
				maxirtemp = irtemp[k];
			}
		}
		for(unsigned k=0; k< irconvol.size(); ++k)
		{
			irconvol[k] = irtemp[k] / maxirtemp;
		}
	}


	convolve(irconvol, vCoef, ir);

	double maxir = *(std::max_element(ir.begin(), ir.end()));
	if(maxir < 0)
	{
		for(unsigned i = 0; i < ir.size(); ++i)
		{
			ir[i] = abs(ir[i]);
		}
		maxir = *(std::max_element(ir.begin(), ir.end()));
	}
	rIRout.resize(ir.size(), 0.);
	copy(ir.begin(), ir.end(), rIRout.begin());
	for(unsigned i = 0; i < ir.size(); ++i)
	{
		irout[i] = (ir[i]) / maxir;
	}

	unsigned k1 = 0;
	double x1 = 0;
	double x2 = 0;
	double irc = 0.5;// pow(10, -6 / 20.)


	for(unsigned jf = 0; jf < nk - 1; ++jf)
	{
		if((irout[jf + 1] - irc) * (irout[jf] - irc) <= 0)
		{
			if(abs(irout[jf] - irc) < 1E-9)
			{
				k1 = jf;
				x1 = k1;
			}else if (abs(irout[jf + 1] - irc) < 1E-9)
			{

				k1 = jf + 1;
				x1 = k1;
			}else
			{
				double weight = abs(irout[jf] - irc) / abs(irout[jf + 1] - irout[jf]);
				k1 = jf;
				x1 = jf + weight;
			}
			break;
		}
	}

	for(unsigned jf = k1 + 1; jf < nk - 1; ++jf)
	{
		if((irout[jf + 1] - irc) * (irout[jf] - irc) <= 0)
		{
			if(abs(irout[jf + 1] - irc) < 1E-9)
			{
				x2 = jf + 1;
			}else if (abs(irout[jf] - irc) < 1E-9)
			{

				x2 = jf;
			}else
			{
				double weight = abs(irout[jf] - irc) / abs(irout[jf + 1] - irout[jf]);
				x2 = jf + weight;
			}
			break;
		}
	}


	rDzfwhm = vZsampling * (x2 - x1); 
}







int main(int argc, char** argv)
{

	vector<double> impulse(2400, 0.); // vector of 2400 values initialized at 1

//vector<double> filter(9, 1/9.); // vector of 9 values initialized at 1/9.
	vector<double> filter(3,0); // vector of 9 values initialized at 1/9.
	filter[0] = 0.333;
	filter[1] = 0.333;
	filter[2] = 0.333;

	unsigned status = 0;
	double rDz = 0.;
	vector<double> irout;

	NDAAC_ResolIR(100, filter, impulse, status, rDz, irout);

	cout<<"The status after the computation is "<<status<< "  and the calculated resolution is "<<rDz<<endl;

	cout<<"The output function is "<<endl;

	for(unsigned i = 0; i < irout.size(); ++i)
	{
		cout<<irout[i]<<"\t";
		//cout<<impulse[i]<<"\t";
	}
	cout<<endl;
	return 0;
}

File: VerticalResolution/PythonVersion/ResolutionIR.py
#!/usr/bin/env python3

from pylab import *




def NDAAC_ResolIr(vZsampling, vCoef, vIRinp):
    """ Computes the resolution using the FWHM of the impulse response

    ... (add Thierry data)


    Returns Status, 
    Returns dzfwhm: the resolution in the same unit of vZsampling
    Returns irout: the impulse response
    """
    ncoef = len(vCoef)
    nk = len(vIRinp)
    status = 0
    missval = -99
    irout = vIRinp.copy() # we copy the values of vIRinp in irout
    irconvol = vIRinp.copy() # we copy the values of vIRinp in irout
    ir = zeros(nk)  

    if not (ncoef % 2): # the number of coefs is even, we reject
        print("Error, The number of coefficients is even")
        return status, 0, irout

    if ncoef == 1 and vCoef[0] != 1: # We reject when one coeff different from 1
        return status, 0, irout

    if ncoef == 1 and vCoef[0] == 1: # We reject when one coeff different from 1
        status = 1
    else: # at least 3 coefs (number odd) => we calculate the transfer functions 
        nc = (ncoef - 1) // 2
        # we check if the coefficients are symmetric (status 1) or antisymmetric (status 2)
        status = -1
        for ic in range(1, nc + 1): # Python is like IDL, subscript starts at 0
            if abs(vCoef[nc - ic]) < 1E-10 and abs(vCoef[nc + ic]) < 1E-10:
                print("0th coeff")
            elif abs(vCoef[nc - ic] - vCoef[nc + ic]) < 1E-3 * abs(vCoef[nc + ic]):
                if status == 2:
                    status = 0
                    print("error ir, we have a mix of symmetric and antisymmetric")
                    return status, 0, irout
                status = 1
            elif (abs(vCoef[nc - ic] + vCoef[nc + ic]) < 1E-3 * abs(vCoef[nc + ic]) and vCoef[nc] == 0):
                if status == 1:
                    status = 0
                    print("error ir, we have a mix of symmetric and antisymmetric")
                    return status, 0, irout
                status = 2
            else:
                    status = 0
                    print("error ir, we have a mix of symmetric and antisymmetric")
                    return status, 0, irout


#    print "0", irconvol
    if sum(irconvol) == 0 and irconvol.std() == 0:
        irconvol[nk // 2] = 1

    if status == 2:
        irtemp = zeros((len(irconvol)))
        irtemp[0] =irconvol[0]
        for k in range(1, nk):
            irtemp[k] = sum(irconvol[0:k + 1])   
        irtemp = irtemp / max(irtemp)
        irconvol = irtemp

#    print irconvol[1190:1210]

    #print "1", ir
    for j in range(nc, len(irconvol) - nc ):
      # ir[j + nc] = sum(array(irconvol[j-nc :j+nc + 1]) * array(vCoef))
       ir[j] = sum(array(irconvol[j-nc :j+nc + 1]) * array(vCoef)) # We have to do the convolution like this
    
    #print "2", ir





    if max(ir) < 0:
        irout = abs(ir)
    else:
        irout = ir
    maxirout = max(irout)
    irout /= maxirout

#    print irout
    db = -6
    #irc = 10 ** (db / 20.)
    irc = 0.5
#    print irc
    
    k1 = 0
    x1 = 0
    x2 = 0
    
    for k in range(nk - 1):
        if (irout[k + 1] - irc) * (irout[k] - irc) <= 0:
            if abs(irout[k] - irc) < 1E-9:
                k1 = k
                x1 = k1
            elif  abs(irout[k + 1] - irc) < 1E-9:
                k1 = k + 1
                x1 = k1
            else:
                weight = abs(irout[k] - irc) / abs(irout[k + 1] - irout[k])
                k1 = k
                x1 = k + weight
            break

    # print "k1, x1", k1, x1
    for k in range(k1 + 1, nk - 1):
        if (irout[k + 1] - irc) * (irout[k] - irc) <= 0:
            if abs(irout[k + 1] - irc)< 1E-9:
                x2 = k + 1
            elif  abs(irout[k] - irc)< 1E-9:
                x2 = k
            else:
                weight = abs(irout[k] - irc) / abs(irout[k + 1] - irout[k])
                x2 = k + weight
            break
   # print "x2", x2
    dzfwhm = vZsampling * (x2 -x1)
    irout *= maxirout


    return status, dzfwhm, irout










if '__main__' == __name__:
    impulse = zeros((2400))
   # status, dz, impulse =  NDAAC_ResolIr(1, array([1/5,1/5,1/5,1 /5.,1/5]), impulse)
 #   impulse = zeros((2400))
   # print NDAAC_ResolIr(1, 1/9*array([1,1,1,1,1,1,1,1,1]), impulse)

    impulse = zeros((2400))
   # print  NDAAC_ResolIr(1, array([-3,-2,-1,0,1,2,3]) / 28., impulse)
   # print NDAAC_ResolIr(1, ones((25)) / 25, impulse)
 #   status, dz, impulse = NDAAC_ResolIr(7.5, ones((25)) / 25, impulse)
    print("second")
    print(NDAAC_ResolIr(7.5, 1/2*array([-1,0,1]), impulse))
    impulse = zeros((2400))
    print(NDAAC_ResolIr(7.5 * 25, 1/2*array([-1,0,1]), impulse))
    






